### ISSUE TITLE
[Android Bootcamp] Project 0 - Review my app

### ISSUE THREAD

**Issue by ZekunWang** [ZekunWang/ToDoApp#1]
My App is complete, please review. 
/cc @codepathreview @codepath

**Comment by nesquena** [comment]
Zekun,

Looks good, this exercise is intended in part to give you an introduction to the general rhythm of this course. The course is entirely project-based with an app being assigned each week and then due the following week. Each project builds on the last to help each engineer  learn the practical elements of Android development and best practices as quickly as possible. We also do a code review for each submitted project once the program begins.

The next step is to continue working on [extensions to your todo app](http://courses.codepath.com/snippets/intro_to_android/prework#heading-5-extending-your-todo) and to **schedule a short 10 minute phone conversation** with us. Navigate to [today's date and onward from there](https://calendar.google.com/calendar/selfsched?sstoken=UURvT01vbzBza1pnfGRlZmF1bHR8NzVjYzBkN2EyZTRiOGRmYjZlZjhmMjA2N2IyMGMwYWU) and choose an available 15-minute slot in the mornings or evenings. Let us know if none of those times work.

With regards to extensions to implement, we are looking for developers with an interest in user experience and product development so be sure to focus on additional functional enhancements and/or improving the user interface of the app.

Once you select a slot, can you make sure to **include the best number** to reach you at to the invite? Look forward to chatting soon!

**Comment by ZekunWang** [comment]
Hi Nathan,

Thanks for the email. Do you only available on Tuesday evenings, Thursday
noons and Friday noons? These time frame does not work for me. I am
available in the evenings of Monday, Wednesday and Friday. I am also
available anytime on weekends.

Thanks,
Zekun Wang

**Comment by nesquena** [comment]
Hi Zekun,

How about tomorrow at 8pm PST?

**Comment by ZekunWang** [comment]
That will work for me. Thanks for the arrangement.

Zekun

**Comment by nesquena** [comment]
Great, invite sent for tomorrow at 8pm.

**Comment by ZekunWang** [comment]
Thanks. Invitation confirmed.

Zekun

**Comment by ZekunWang** [comment]
Hi Nathan,

I am trying to implement the feature of due date for the todo item but have not found anything useful yet. Can you provide tutorials on how to create pop-up calendar when I click a button and pop-up scrolling time picker?

Zekun

**Comment by nesquena** [comment]
Hi Zekun,

Guide on pop-up time picker is here in our guides: http://guides.codepath.com/android/Using-DialogFragment#displaying-date-or-time-picker-dialogs  

I encourage you to check our searchable guides at guides.codepath.com for anything you need while completing the optionals.

**Comment by ZekunWang** [comment]
Hi Nathan,

I added some features and changes. Please review my app. I am still confused about adding Dialog Fragment, changing menu design and adding notification for the due setting even after reading the topics from searching guides.codepath.com . Do you have any video showing how to do these step by step?

Zekun

**Comment by nesquena** [comment]
Hi Zekun,

> I am still confused about adding Dialog Fragment

Do you have any specific issues I can help with. You may have already seen but presenting a [custom dialog fragment is outlined here](http://guides.codepath.com/android/Using-DialogFragment#custom-view). 

> changing menu design 

Do you mean tweaking the [top action bar](http://guides.codepath.com/android/Extended-ActionBar-Guide#custom-actionbar-styles)?

> adding notification for the due setting

This is going to be more complicated and involves using a background service and [alarm manager](http://guides.codepath.com/android/Starting-Background-Services#using-with-alarmmanager-for-periodic-tasks)

**Comment by ZekunWang** [comment]
Hi Nathan,

I am trying to use DialogFragment to implement editing list item. But
android studio is complaining about showing datepicker and timepicker
because it cannot resolve "getSupportFragmentManager()". I think it comes
from extending FragmentActivity but I have to extend DialogFragment. Does
this mean I cannot let user to pick time and date in a DiaglogFragment edit
mode?

Thanks,
Zekun Wang

On Fri, Jun 17, 2016 at 12:10 AM, Nathan Esquenazi <notifications@github.com

**Comment by nesquena** [comment]
Your activity should extend `FragmentActivity` or `AppCompatActivity`:

```
class MyActivity extends AppCompatActivity {
  // ....
}
```

and then you should have a separate dialog fragment object in a separate Java class:

```
class MyDatePickerFragment extends DialogFragment {
  // ...
}
```

then the `getSupportFragmentManager()` is called from within the activity extending `AppCompatActivity`. You would implement `implements DatePickerDialog.OnDateSetListener` inside the activity as well. 

Does that help clarify? Anywhere you see the fragment manager should be inside the activity and then separately there's a dialogfragment class.

**Comment by ZekunWang** [comment]
Yes, I can understand. This is want I did when using activity to edit
items. But if I am using a DialogFragment to edit items, the context should
be a fragment rather than an activity. How can I pop up a date picker when
I click a button in DialogFragment?

Zekun

**Comment by nesquena** [comment]
> Yes, I can understand. This is want I did when using activity to edit
> items. But if I am using a DialogFragment to edit items, the context should
> be a fragment rather than an activity. How can I pop up a date picker when
> I click a button in DialogFragment?

See http://guides.codepath.com/android/Using-DialogFragment#passing-data-to-parent-fragment for a breakdown on how to launch a dialog from a dialog. It's a good learning, common use case.

**Comment by ZekunWang** [comment]
Hi Nathan,

I read this already. However, the requirement is to support edit item in
dialog fragment. So when the dialog fragment pops up and let me edit the
item, I should be able to click the button in the dialog and it will pop up
a date picker dialog. When I choose a date and click "ok", it will return
to the edit dialog and let me do further changes. Is it possible to that?

Thanks,
Zekun

On Tue, Jun 21, 2016 at 12:31 PM, Nathan Esquenazi <notifications@github.com

**Comment by nesquena** [comment]
That's that this guide demonstrates: http://guides.codepath.com/android/Using-DialogFragment#passing-data-to-parent-fragment with the date being selected in a dialog and then being sent back to the other dialog fragment. If you are finding this story too frustrating to implement, you can move on from this optional too, we will be covering again in more detail as part of the course.

**Comment by ZekunWang** [comment]
Hi Nathan,

Thanks for your help. I am still confused so I will skip this. I have
submit new updates of ToDoApp. Please review the app.

Thanks,
Zekun

On Wed, Jun 22, 2016 at 12:56 AM, Nathan Esquenazi <notifications@github.com
